### PR TITLE
Filament Width Sensor Support

### PR DESCRIPTION
**Pull request by nopbit**
Support added for filament diameter sensor. Sample configuration parameters added to example-extras.cfg, M407 - Filament Width command added to "Report the current measured filament width to the host". If reading is not in range between + - max_difference nominal filament width applied. If filament diameter is smaller than 0.5mm it clears all readings. I tested it on my printer and it seems to work. Waiting for your comments. Thanks for you support Kevin

Regards
Mustafa

### CODE REVIEW

**Comment by glowtape** [comment]
What sensor are you using for this?

**Comment by nopbit** [comment]
I have built one such sensors on thingiverse https://www.thingiverse.com/thing:704897, it is based on this design https://www.thingiverse.com/thing:454584 I was using it with Marlin. Sensor is generating dc voltage which is equal to filament width. For example 1.75V for 1.75mm.

Regards,

File: klippy/extras/filament_width_sensor.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                        percentage = round(self.nominal_filament_dia / filament_width * 100)
+                        self.gcode.cmd_M221({'S': percentage})
+                    else:
+                        self.gcode.cmd_M221({'S': 100})
+        else:
+            self.gcode.cmd_M221({'S': 100})
+            self.filament_array = []
+        return eventtime + 1
+
+    def cmd_M407(self, params):
```
[the remark this answers is not in the document]
Comment (nopbit) [reply klippy/extras/filament_width_sensor.py]: I believe M407 is standard G-Code, i was using it in Marlin. [(Reprap Wiki for M407)](https://reprap.org/wiki/G-code#M407:_Display_filament_diameter) I will change it with your suggestion to make it human readable and user-friendly.

File: klippy/extras/filament_width_sensor.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        # extrude factor updating
+        self.extrude_factor_update_timer = self.reactor.register_timer(
+            self.extrude_factor_update_event)
+
+    # Initialization
+    def handle_ready(self):
+        # Load printer objects
+        self.gcode = self.printer.lookup_object('gcode')
+        self.toolhead = self.printer.lookup_object('toolhead')
+        self.heater_bed = self.printer.lookup_object('heater_bed', None)
```
[the remark this answers is not in the document]
Comment (nopbit) [reply klippy/extras/filament_width_sensor.py]: Its my mistake during copy and paste, removed

File: klippy/extras/filament_width_sensor.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ADC_SAMPLE_COUNT = 8
+MEASUREMENT_INTERVAL_MM = 10
+
+class FilamentWidthSensor:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.reactor = self.printer.get_reactor()
+        self.pin = config.get('pin')
+        self.nominal_filament_dia = float(config.get('default_nominal_filament_dia'))
+        self.measurement_delay_mm = float(config.get('measurement_delay_cm')) * 10
```
[the remark this answers is not in the document]
Comment (nopbit) [reply klippy/extras/filament_width_sensor.py]: All units converted to mm in config file

File: klippy/extras/filament_width_sensor.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        if self.lastFilamentWidthReading > 0.5:
+            if len(self.filament_array) > 0:
+                pending_position = self.filament_array[0][0]
+                if pending_position <= last_epos:
+                    # Get first item in filament_array queue
+                    item = self.filament_array.pop(0)
+                    filament_width = item[1]
+                    if (filament_width <= self.max_diameter) and (filament_width >= self.min_diameter):
+                        percentage = round(self.nominal_filament_dia / filament_width * 100)
+                        self.gcode.cmd_M221({'S': percentage})
```
[the remark this answers is not in the document]
Comment (nopbit) [reply klippy/extras/filament_width_sensor.py]: Replaced with gcode.run_script. Thanks

**Comment by KevinOConnor** [comment]
Thanks.  In general, it looks okay to me.

What's the name of the hardware you are using to sense the width?  I suspect there might be other types of filament width sensors, so I wonder if we can name this one to be more precise (maybe something like adc_shadow_filament_sensor)?  @Arksine - I know you were working on a "laser mouse" type sensor - does that do filament width sensing?

As a minor comment, I think it would be preferable to spell out "diameter" in the config file.  It would also be preferable for the config documentation to specify the units (it's not clear to me what measurement_delay is) and what the defaults are explicitly (like the other config documentation does).

Cheers,
-Kevin

**Comment by nopbit** [comment]
Hello Kevin,

> What's the name of the hardware you are using to sense the width? I suspect there might be other types of filament width sensors, so I wonder if we can name this one to be more precise (maybe something like adc_shadow_filament_sensor)? 

I am using a sensor which is using a TSL1401CL (a linear sensor array consists of a 128 × 1 array of photodiodes) which is driven by an arduino pro micro. When filament is passing between LED and linear array sensor a shadow fall on sensor. According to this information arduino pro micro generate an analog voltage output which is equal to filament width. If filament width is 1.60mm it generates 1.60volt, if 3mm it generates 3.0V at output. This is something universal actually. Any sensor which generates this analog voltage output can be used with my code.  I used below code to convert adc reading to Voltage.

`self.lastFilamentWidthReading = round (read_value * 5, 2)`

> As a minor comment, I think it would be preferable to spell out "diameter" in the config file. It would also be preferable for the config documentation to specify the units (it's not clear to me what measurement_delay is) and what the defaults are explicitly (like the other config documentation does).
> 
O.K. I will change default_nominal_filament_dia -> default_nominal_filament_diameter

I added comments in example configuration but if you think they are not enough or clear i can also create a documentation file. 

#   The distance from sensor to the melting chamber as mm
#measurement_delay 100

Filament sensor is not closer or inside hotend so I am using measurement_delay parameter for making my readings valid. For example if i place the filament sensor 150mm before the melting chamber my first readings position will be [150mm, x.xx diameter]. Maybe I can rename 'measurement_delay' to "**distance_to_melting_chamber**". I believe it will be more clear like that to everyone :)


I am also planing to add more commands like "DISABLE_FILAMENT_WIDTH_SENSOR, ENABLE_FILAMENT_WIDTH_SENSOR", I can also add support to "pause print" if filament width is smaller than a certain value. 

Thanks,
Best regards
Mustafa




> Cheers,
> -Kevin

**Comment by KevinOConnor** [comment]
Thanks.  It looks fine to me.  The only thing I'm not sure about is the module name.  Does the sensor you designed have a "name"?  Could we call this module something like the "foomatic_filament_sensor" (where "foomatic" is the name you use to describe this particular sensor)?  The minor concern I have is that individuals with different filament sensors will see this module and spend time trying to configure it, without realizing it is for different hardware then they own.

Otherwise, everything looks okay to me and I'm happy to merge.

-Kevin

**Comment by nopbit** [comment]
Thank you very much for your comments and support Kevin, I will rename files and add "tsl1401cl_" as prefix. TSL1401CL is the name of the linear array sensor that is connected to arduino.

Regards
Mustafa

**Comment by KevinOConnor** [comment]
Thanks!

-Kevin

**Comment by jdlongenecker** [comment]
@nopbit how has this been working? I'm interested in trying it out if you've had success with it.

**Comment by nopbit** [comment]
Dear @jdlongenecker, I have been using it for a long time with no problem. But if you are using a branded filament witch nice Quality Control, it is not necessary. I am using it with DIY no branded filaments from my my friend.

**Comment by jdlongenecker** [comment]
Got it! The thought of not having to measure filament diameter before printing was convenient to me, but ultimately it's a minor quality of life improvement. That's awesome that you're able to get DIY filament!

**Comment by carver0510** [comment]
Would it be possible to use the Sensor also for filament run out detection?

Because when there will be no filament seen on the sensor, there could be a signal triggerin the runout?

**Comment by nopbit** [comment]
Yes it is possible to use the sensor for filament run out detection but we need to add some code in plugin file.

**Comment by carver0510** [comment]
This would be an awesome feature. Maybe there could be some trouble with clear filaments, but it would be very nice if somebody could add some code.
At the moment I am not this in coding :/

**Pull request merged**